### ISSUE TITLE
Compound concatenation vs Long-form code (aka The array vs the stream)

### ISSUE THREAD

**Issue by footloosejava** [iotaledger/iri#603]
Since I am on a tear with Rants et Al.  ... I thought I'd throw this one out for everyone's Thursday Hump-day enjoyment and wonder. 

Please don't let me distract you from more important matters - this is strictly for your coding enjoyment and life-changing reflections .... 

The target of our rambling this instant is **StateDiff.java** where we want to concatenate all the entries byte value into a single byte[]. The source array is a `Map<Hash, Long>` , 

As it stands in the dev branch:

```
public byte[] bytes() {
    return state.entrySet().parallelStream()
        .map(entry -> ArrayUtils.addAll(entry.getKey().bytes(), Serializer.serialize(entry.getValue())))
        .reduce(ArrayUtils::addAll)
        .orElse(new byte[0]);
}
```

What is really going on there? If we unroll that, it happens that:

1. Each entry is turned into a byte array (new byte[]) by  `ArrayUtils.addAll`
2. Each result of (1) is then also concatenated by ` ArrayUtils.addAll' to the previous value.

For ten entries, we will produce 10 new byte[] from (1) and then 9 more byte[] from (2). So how many byte[] depends on the size of the map, but it is basically `new byte[] count = (2 x Map.size) -1;

If the `Map` has 10 entries, we produce 19 new byte[] in the process. If it has one byte[], then we produce 1 byte[] in the process.

There is probably a negligible time involved (for small sets with small byte[] entries) and probably a negligible impact on performance time, garbage collection, etc. From that I would conclude that for small Maps with small byte[] contents conversion, this is a clear, bug-free way to produce the desired output.

However, there is a faster, leaner way that needs more testing but would always be efficient:

```
public byte[] bytes() {
    int size = state.size();
    if (size == 0) {
        return new byte[0];
    }

    byte[] temp = new byte[size * (Hash.SIZE_IN_BYTES + Long.BYTES)];
    int index = 0;
    for (Map.Entry<Hash, Long> entry : state.entrySet()) {

        byte[] key = entry.getKey().bytes();
        System.arraycopy(key, 0, temp, index, key.length);
        index += key.length;

        byte[] value = Serializer.serialize(entry.getValue());
        System.arraycopy(value, 0, temp, index, value.length);
        index += value.length;
    }
    return temp;
}
```

We make no more than **one** new byte[], the correct size, and copy the contents in. 

I suppose this could be compute-paralleled if the order of the contents didn't matter. To do that, a stream could be used to feed the Map entries -along with an unused Integer index- to a routine that would array-copy the bytes from an entry into an unused slot. The result is probably the same as the original since the parrallel operation does not guarantee order.

**Unneeded performance optimization?** 

**Betterment of a routine to minimize time and garbage in the VM?**

I guess it all comes down to: What size is the map expected to be? How many bytes in each byte[] for keys and values. For large maps and big byte counts ... there could be an advantage to the long form copy.

☕️  + 🍰  = 😄 

I hope, whatever your contributions to whatever open-source projects you are following ... you have lots of coffee and cake.  The end.

**Comment by footloosejava** [comment]
I don't see any point (or urgency) in keeping this open. I am closing this issue.

**Comment by GalRogozinski** [comment]
Hey,
I don't know if it will makes you feel better. Before I joined IF I had an assignment to perform a change in the RocksDB persistence provider.
I noticed the same issue and performed a change that is similar to yours. It can still be found on my home PC. Due to priority changes this was left behind.

We do need to perform this change still. The reason I am not doing it at this very moment because we are currently at a code freeze.

Anyhows, I am reopening this issue because it is a good issue.
